### PR TITLE
Only give actors ScriptTriggers on scripted maps in official mods

### PR DESCRIPTION
**Pull request by reaperrr**
This should reduce the amount of some `INotify*` calls (especially `INotifyKilled`) substantially during normal MP/skirmish games, reducing worst-case peaks.

### CODE REVIEW

File: mods/cnc/rules/campaign-maprules.yaml [hunk 1]
```diff
@@ -40,6 +42,21 @@ Player:
 		Name: Campaign Player AI
 		Type: campaign
 
+^ExistsInWorld:
+	ScriptTriggers:
+
+^CivBuildingHusk:
```
Comment (pchote) [inline mods/cnc/rules/campaign-maprules.yaml:48]: Aren't these already handled by `^ExistsInWorld`?

Comment (reaperrr) [reply mods/cnc/rules/campaign-maprules.yaml:48]: No, cnc `^CivBuildingHusk` doesn't inherit that since most of the stuff on ExistsInWorld is redundant for them.

Same for the others below.

**Comment by abcdefg30** [comment]
There is perf to gain here, but is  it worth breaking basically every custom map using Lua?

**Comment by abcdefg30** [comment]
Actually, I guess it is possible to write an update rule for this, but I'm unsure about the complexity of that.

**Comment by pchote** [comment]
```
If (LuaScript trait used)
    yield "you may need to fix ScriptTriggers"
```

**Comment by reaperrr** [comment]
Did some more profiling of `Health.InflictDamage`.
Result: All cases where some projectile, `Explodes` or similar shows up in perf.log on the RA shellmap, it's indeed almost entirely due to INotify* calls to `ScriptTriggers`.

I'll update this later.

**Comment by reaperrr** [comment]
OK, after some more testing this only costs perf on maps with actual Lua scripts, not in vanilla skirmish/MP games, so I don't think this PR is worth the hassle.

**Comment by pchote** [comment]
The `INotifyOtherProduction` implementation in `ScriptTriggers` means that every non-decoration actor on the map gets pinged everytime anybody produces anything. This is really wasteful and I would be very suprised if this didn't contribute at least somewhat to the performance issues in large team games.

If we dont want to go ahead with the changes here, then a less invasive solution to this specific case could be to add a `ScriptTriggersManager` trait to the world actor that implements `INotifyOtherProduction` (or maybe plumb it via `ActorMap`?) and expose an event that the individual `ScriptTrigger` trait instances can attach a callback to *only if they need it* (which is almost never).

**Comment by reaperrr** [comment]
I just ran multiple tests of 6v6 bots team games on bleed RA, not sure though how much different things might be with human players, though.

`ScriptTriggers` didn't show up in perf.log even once, even with threshold lowered to 0.001 ms.
The only things showing up a ton (several times per tick) were `ModularBot`, `FrozenActorLayer` and `Shroud`, and less frequent but sometimes several times at once the usual suspects like `Mobile`, `AttackTurreted`, `FindAndDeliverResources`, and of course `AttackMove`, `Move` and its `*Half` childs.
So at least on the AI side, the first 3 are responsible for virtually all perf issues not related to pathfinding, everything else appears to be negligible in comparison, especially `ScriptTriggers`, unless it's somehow different for human players.

**Comment by pchote** [comment]
Need to be careful there: the performance overhead from notifications is counted as part of the trait making the notifications rather than the trait being notified. So e.g. `INotifyOtherProduction` would be counted under `Production` etc.

**Comment by reaperrr** [comment]
Good point and `Production.DoProduction` as a whole is indeed on the expensive side (wonder why the trait doesn't show up in perf.log with normal logging), but to put into perspective how much of that seems to be because of `INotifyOtherProduction` to ScriptTriggers:

>   6748 sys ticks [10105] DoProduction - FrameEndTask other stuff, tent: Production
   528 sys ticks [10105] DoProduction - notify e3: e3->ScriptTriggers

This was (by far) the largest cost I was able to measure for the notification in another 6v6 AI game. I set up my logging to only show when exceeding 24 stopwatch ticks, and most of the time it apparently doesn't even do that.

I then did one last run measuring each part of the 'other stuff' separately.
tl;dr: The FrameEndTask cost is almost entirely down to the creation of the new actor. It's usually in the 4K-7K stopwatch ticks, or 0.4-0.7ms ballpark on my machine, but occasionally (rarely) spikes to anywhere up to 3.8ms, which could of course easily end up at multiples of that on slow CPUs.

But I don't see production spawning enough actors in the same tick to become a real bottleneck (normally, at least).
I'd expect `Move` - either directly or as child activity - to still be behind the bulk of perf issues in large games.

Anyway, the `INotify*` pings are indeed not free, but very cheap in the grand scheme of things. The  `ScriptTriggersManager` approach sounds like too much work for insignificant gain. I'd say it's either reviving this PR or leaving things as they are. We can likely win far more by going through the actor creation process and expensive traits again to see what we can do there.

**Comment by anvilvapre** [comment]
> Anyway, the INotify* pings are indeed not free, but very cheap in the grand scheme of things. 

I looked at ScriptTriggers in the past and came to the same conclusion. A lot of callbacks. 

I wondered if perhaps there could be gain in moving more of the code in Lua to avoid having to create the Lua context for each call. But from what i remember that wasn't easily possible.

And assuming that triggers are now fired for any actor. Not sure. I wondered if you could allow Lua users - in a new API - to register per actor what events they want to receive. Or per actor mask. I.e. only be notified of TickIdle for AI actors. They would register which triggers they want to receive once they are notified an actor was created.
